### DISCUSSION TITLE
What happens to noun and verb during a program alarm?

### DISCUSSION THREAD

**Discussion by jumpjack** [virtualagc/virtualagc#1166] [Q&A] [answered]
During 1202 alarm I see verb and noun changing to 50 and 90, but I cannot understand if they are user input or computer requests.

 - I [read](https://pulsar.sternwarte.uni-erlangen.de/wilms/computers/apollo.html) that verb 50 means "Please perform", i.e. it's an AGC request to the operator.
 - I also [read](https://www.discovermagazine.com/the-sciences/how-verbs-and-nouns-got-apollo-to-the-moon) that verb 50 means user request to AGC
- I also read that noun 90 does not even exist

How can I understand from [the source](
https://github.com/virtualagc/virtualagc/blob/89c80bbe7f1b5318acd17f0dd4725105c41a12b3/Comanche045/EXECUTIVE.agc#L188) if something is written to verb/noun or expected?

**Comment by rburkey2005** [comment]
It would help if you said where you read that noun 90 does not exist.  Otherwise it's a bit tough to explain why it may have said that.

There's certainly a noun 90 in Luminary 99, because you can see it [in the source code (look for the three lines beginning at "012877,000373" in column 1)](https://www.ibiblio.org/apollo/listings/Luminary099/PINBALL_NOUN_TABLES.agc.html).  As far as I can see, it's present in every version of Luminary available to us.

I'm ignorant, so as I promised, I'm going to notify the mailing list and see if anybody who knows more might want to respond.  But here are my immediate thoughts, just in case.  What the display of 50 as the verb is doing, I think, is asking the astronaut to do something.  The display of 90 as the noun is telling the astronaut what it wants done, which (as you can see from the link I gave you above) is to enter Y, Y DOT, and PSI. 

As far as following the source is concerned, It's not really something I'm able to do myself with any degree of confidence, so to me it seems really hard.  The EXECUTIVE (which is the source module your question linked) is a generic module that manages the various jobs running simultaneously in this multi-tasking operating system, so it's not going to give you any specifics as to what any given job is doing, though it's true in this case that 01202 is an error in the executive.  What happens when there's an alarm is typically that a subroutine like BAILOUT (there are several alternatives) is called, so the question is really what those subroutines might be doing, and not what the EXECUTIVE is doing.  In the case of a 1202, the EXECUTIVE calls [BAILOUT1 in ALARM AND ABORT](https://www.ibiblio.org/apollo/listings/Luminary099/ALARM_AND_ABORT.agc.html).

Looking at the software flowcharts is likely going to be easier than trying to follow the source code.  We don't have the flowcharts for Luminary 99, but we do have Luminary 173 flowcharts, and the aspects of the code you're asking about aren't likely to have changed much.  For example, [the flowchart for BAILOUT1 specifically is here](https://www.ibiblio.org/apollo/Documents/j2-80-E-2471-REV2-VOL1_text.pdf#page=143).  But bear in mind that when I say the flowcharts are _easier_, that shouldn't be taken to mean that they're _easy_.  If you do choose to look at the flowcharts, they're gigantic and are likely to bog your browser down to a crawl, so I'd suggest downloading both volumes ([the 2nd volume is here](https://www.ibiblio.org/apollo/Documents/j2-80-E-2471-REV2-VOL2_text.pdf)) and using a PDF viewer directly on your local computer rather than using a browser to look at them on the website.

**Comment by indy91** [comment]
To look at the program alarms you do V05 N09, not V50 N90. So researching V50 and N90 is going to mislead you a bit. I don't know why the telemetry says 50 and 90 though. Maybe because they actually typed in V5 N9 not V05 N09 and that appears in a wrong way on the processed telemetry, or something like that.

**Reply by rburkey2005** [reply]
Yes, that makes a lot more sense.

**Comment by thewonderidiot** [comment]
I hadn't thought of it, but I 100% agree with indy91's theory that they simply typed V5N9E instead of V05N05E. In telemetry that would look like "5blank 9blank", which might have been implicitly changed to "50 90" in whatever telemetry system they were using.

**Reply by ehdorrii** [reply]
I've looked at the telemetry "listings" that you are working from. I don't believe that the downlinked data was wrong, nor was the storage of the downlinked data. I believe that the program that was written to _format and print_ the data was at fault. If you look at the data, any time there is a single digit in the "NOUN DISPLAYED" or the "VERB DISPLAYED" column (for example " 6") note that the leading zero is not shown -- the leading zero would in fact be shown on the DSKY -- it would be _displayed_ as "06". Note also that a _decimal point_ is printed, so it is given as " 6.", not simply " 6". That tells me that whoever wrote the program that formatted and printed the data, used a numeric format to print the entire field -- that it wasn't formatted character-by-character as the R1, R2 and R3 displays were. 

Remember that this all happened before the C language was even invented. The programmer would have used something like RPG, a language specifically designed to format and print data, with built-in number formatting. The problem with formatting the R1 through R3 fields as numbers would have been immediately apparent, since numeric formatting would not prepend a plus or minus sign, and would drop any leading zeros. But the Major-Mode, Verb and Noun fields wouldn't necessarily be obviously wrong until you started looking at them closely. Most of the time they would be okay.

Since this is a numeric field, there is just no way to print six-with-a-space-after-it -- "6 ." -- that simply isn't a valid numeric output. The programmer could have formatted these fields character-by-character as the R1-R3 fields are, but didn't. It seems like an obvious bug, but most bugs are obvious in hindsight. It may be that anyplace a "blank" was encountered, it was converted to "zero", in the belief (by the programmer) that blanks would only appear to the left of significant digits.

**Comment by jumpjack** [comment]
any theory is good without supporting documentation.

**Comment by thewonderidiot** [comment]
Okay, you're welcome. I'm all ears if you come up with a different explanation, but as indy91 was getting at, V50 N90 is not a valid combination. It will never come up automatically, and manually entering it won't do anything -- especially not pull up the program alarms, as is shown in the telemetry. V5N9E will end up with the 5 and 9 in the correct positions and the program alarms displayed, leaving only the 0's unexplained.

**Comment by jumpjack** [comment]
>  V50 N90 is not a valid combination. It will never come up automatically, and manually entering it won't do anything -- especially not pull up the program alarms, as is shown in the telemetry

Yes, this is what my question started from.

Apparently the V5N9E sequence and the telemetry bug is the only explanation, so now I wonder if it is possibile to locate the exact lines of the source code which "build" the telemetry and send it to ground, just to figure out if it is an LGC bug or a ground SW bug.

**Reply by rburkey2005** [reply]
This is the best kind of question ... the kind I know the answer to.  :smiley: 

The AGC software does not "build" any data that it sends to the ground.  There is simply a list of memory locations (maintained in DOWNLINK_LISTS).  The AGC software (DOWN-TELEMETRY PROGRAM) fetches the data from those memory locations and transmits it, as-is.

In the case of the DSKY display, that data happens to be in the 12-word variable DSPTAB.  DSPTAB is precisely the same data which the AGC software sends to the DSKY.  So whether or not it can be in error is a nice philosophical problem, to which I suspect the answer would have to be "no".

But at any given time, DSPTAB is only going to reflect what's actually displayed on the DSKY at that specific time.  So if the astronaut has keyed in VERB 5 (rather than VERB 05), DSPTAB is going to contain something that represents 5 (blank) rather than 0 5.  

Here's a screenshot of the (simulated) DSKY display _and_ the telemetry if you just start it up and do V5N9E:

![DSKY-V5N9E](https://user-images.githubusercontent.com/11841400/154801706-124976ca-d452-4cb7-8a7f-a8539eb24c0d.png)

You can decode the telemetry's "DSPTB" yourself to see whether it contains something reflective of 50 90, 5(blank) 9(blank), or 05 09.

**Reply by rburkey2005** [reply]
Darn it!  I was curious enough to decode it myself.  According to PINBALL_GAME_BUTTONS_AND_LIGHTS, the verb-area on the DSKY corresponds to word 9 (starting from 0) of DSPTAB and the noun-area corresponds to word 8.  So both are in DSPTB+8 of the telemetry, which has the value (octal) 0574005700.

Decoding 0574005700, the 5-bit relay codes for the individual digits in DSPTB+8 turn out to be:

- 1st digit of NOUN = 11111, or 9
- 2nd digit of NOUN = 00000, or BLANK
- 1st digit of VERB = 11110 or 5
- 2nd digit of VERB = 00000 or BLANK

Exactly as advertised.

**Reply by rburkey2005** [reply]
Perhaps the confusion comes from the fact that the DSKY's 5-bit relay code for a digit of "0" is 10101, whereas the 5-bit relay code for a BLANK digit is 00000.  It's hard to believe that whatever software was being used on the ground for interpreting the telemetry didn't know that, but relatively speaking, it would be easy to mistake a relay code of 00000 as meaning a digit of "0".

**Comment by jumpjack** [comment, marked as answer]
Indeed this is what "source code builds the telemetry" means:  how does the source code turn keypresses V + 5 + N + 9 + E into bits that it sends to ground?

I assume what you say about how AGC works is right... but I didn't understand anything. :-)
Anyway I [found ](https://github.com/virtualagc/virtualagc/blob/b0af9b18420adf6f318bcbad3205973acec36438/Luminary099/PINBALL_GAME__BUTTONS_AND_LIGHTS.agc#L446) the values you mentioned:


- **BLANK	00000**
- **0	10101**
- 1	00011
- 2	11001
- 3	11011
- 4	01111
- 5	11110
- 6	11100
- 7	10011
- 8	11101
- 9	11111



So the bug was on ground SW: the data "5-blank" and   "9-blank" (11110-00000 and  11111-00000)  sent by LGC was interpreted and logged as "5-0"  and  "9-0", due to a misunderstanding between "blank" and "zero" (00000 vs 10101).

Fortunately it was not a critical telemetry.

A possible confirmation that the logged  V50N90 (actually V5N9) was an astronaut command and not an AGC output comes from timing:

![image](https://user-images.githubusercontent.com/1620953/154807812-3204299c-bbcc-423a-9179-d06daaa222ce.png)


The appearance of alarm code in R1 and of V50 and N90 in the log is not immediate, it occurs some seconds after the alarm is detected by AGC... just the time needed to type V5N9  , needed to ask AGC "What does PROG ALARM light refer to?"

**Reply by rburkey2005** [reply]
> Indeed this is what "source code builds the telemetry" means: how does the source code turn keypresses V + 5 + N + 9 + E into bits that it sends to ground?

I beg to differ.  That has nothing to do with the telemetry, and pertains solely to how the AGC communicates with the DSKY.  The telemetry is just like a wiretap on the AGC-to-DSKY communication.  Other than the fact that the telemetry listens in on that conversation between the AGC and DSKY, nothing whatever about the process of communicating between the AGC and DSKY has anything to do with anything the telemetry is doing.

As for how the AGC's software receives keystrokes entered by the astronaut on the DSKY and (among other things) converts them into relay settings that it echos back to the DSKY for display, you're asking how the PINBALL_GAME_BUTTONS_AND_LIGHTS program works.  That program is 4000 lines long, so it's difficult to answer in a few words.  

In general, when a keystroke from the DSKY appears, there's a hardware interrupt that's serviced in INTERRUPT_LEAD_INS.  It calls a subroutine called KEYRUPT1 (in KEYRUPT,_UPRUPT) which saves the incoming data and starts up CHARIN (in PINBALL_GAME_BUTTONS_AND_LIGHTS) which continues working with the data.  Perhaps that's the place for you to look.

**Reply by ehdorrii** [reply]
IMHO looking anywhere for "errors in telemetry" is a waste of time. As I mentioned in my earlier reply, the issue is almost certainly in the program that someone wrote to _format and print_ the received data.

**Comment by rburkey2005** [comment]
In fact, if the astronaut V6, then 6 is what would be displayed, not 06.  Not that that invalidates the rest of what you say.Sent on my Samsung Galaxy A51.
-------- Original message --------From: Gene Dorr ***@***.***> Date: 3/10/22  2:47 PM  (GMT-06:00) To: virtualagc/virtualagc ***@***.***> Cc: Ronald Burkey ***@***.***>, Comment ***@***.***> Subject: Re: [virtualagc/virtualagc] What happens to noun and verb during a
  program alarm? (Discussion #1166) 
I've looked at the telemetry "listings" that you are working from. I don't believe that the downlinked data was wrong, nor was the storage of the downlinked data. I believe that the program that was written to format and print the data was at fault. If you look at the data, any time there is a single digit in the "NOUN DISPLAYED" or the "VERB DISPLAYED" column (for example " 6") note that the leading zero is not shown -- the leading zero would in fact be shown on the DSKY -- it would be displayed as "06". Note also that a decimal point is printed, so it is given as " 6.", not simply " 6". That tells me that whoever wrote the program that formatted and printed the data, used a numeric format to print the entire field -- that it wasn't formatted character-by-character as the R1, R2 and R3 displays were.
Remember that this all happened before the C language was even invented. The programmer would have used something like RPG, a language specifically designed to format and print data, with built-in number formatting. The problem with formatting the R1 through R3 fields as numbers would have been immediately apparent, since numeric formatting would not prepend a plus or minus sign, and would drop any leading zeros. But the Major-Mode, Verb and Noun fields wouldn't necessarily be obviously wrong until you started looking at them closely. Most of the time they would be okay.
Since this is a numeric field, there is just no way to print six-with-a-space-after-it -- "6 ." -- that simply isn't a valid numeric output. The programmer could have formatted these fields character-by-character as the R1-R3 fields are, but didn't. It seems like an obvious bug, but most bugs are obvious in hindsight. It may be that anyplace a "blank" was encountered, it was converted to "zero", in the belief (by the programmer) that blanks would only appear to the left of significant digits.

**Reply by ehdorrii** [reply]
Yes, you're right. I guess I left out some pertinent details. If the astronaut keyed **V 0 6**, then 06 would be in the display, and " 6." would appear in the formatted listing. If the astronaut keyed **V 6**, then 6_ would be in the display (the underscore indicating a blanked digit), and "60." would appear in the listing. The problem, of course, is that the latter would be indistinguishable from the situation where the astronaut keyed in V 6 0. Yes, a program bug, but not in the AGC, not in the downlinked data, not in the stored data -- the bug is in the program that formatted and printed the stored telemetry. Something to bedevil people looking at the data after the fact.
